addons/mail/static/src/model/model_core.js

```diff
@@ -2,134 +2,132 @@
+export function patchFields(modelName, fields) {
+    const definition = registry[modelName];
+    if (!definition) {
+        throw new Error(`Model "${modelName}" must be registered before patched.`);
+    }
+    for (const [fieldName, field] of Object.entries(fields)) {
+        if (definition.fields.hasOwnProperty(fieldName)) {
+            throw new Error(`Error patching "${modelName}/${fieldName}": patch to an existing field is not allowed.`);
+        }
+        definition.fields[fieldName] = field;
+    }
 }
 
 /**
- * Register a patch for fields in model.
+ * Applies the provided function to the identifying fields of the model
+ * specified by the `modelName`.
  *
- * @param {string} modelName
- * @param {string} patchName
- * @param {Object} patch
+ * @param {string} modelName The name of the model to which to apply the patch.
+ * @param {function} patch The function to be applied on the identifying fields.
  */
-function registerFieldPatchModel(modelName, patchName, patch) {
-    _registerPatchModel(modelName, patchName, patch, { type: 'field' });
+export function patchIdentifyingFields(modelName, patch) {
+    const definition = registry[modelName];
+    if (!definition) {
+        throw new Error(`Model "${modelName}" must be registered before patched.`);
+    }
+    patch(definition.identifyingFields);
 }
 
 /**
- * Register a patch for identifyignFields in model.
+ * Adds or overrides the provided methods to the model specified by the
+ * `modelName`.
  *
- * @param {string} modelName
- * @param {string} patchName
- * @param {function} patch
+ * @param {string} modelName The name of the model to which to apply the patch.
+ * @param {Object} methods Methods to be added or overriden.
  */
-function registerIdentifyingFieldsPatch(modelName, patchName, patch) {
-    _registerPatchModel(modelName, patchName, patch, { type: 'identifyingFields' });
+export function patchModelMethods(modelName, methods) {
+    const definition = registry[modelName];
+    if (!definition) {
+        throw new Error(`Model "${modelName}" must be registered before patched.`);
+    }
+    if (!definition.modelMethods) {
+        definition.modelMethods = {};
+    }
+    for (const [methodName, method] of Object.entries(methods)) {
+        if (typeof method !== 'function') {
+            throw new Error(`Cannot patch model method on model "${modelName}": ${methodName} is not a function.`);
+        }
+        if (!definition.modelMethods[methodName]) {
+            definition.modelMethods[methodName] = method;
+            continue;
+        }
+        const originalMethod = definition.modelMethods[methodName];
+        definition.modelMethods[methodName] = function(...args) {
+            const previousSuper = this._super;
+            this._super = originalMethod;
+            const ret = method.call(this, ...args);
+            this._super = previousSuper;
+            return ret;
+        };
```

Comment:
Could be nice to comment why this is useful. I know it's for making sure calling `_super` climbs up the parent relations properly, but it can be good to comment so it's clearer for other readers.